common/src/main/java/com/microsoft/identity/common/adal/internal/tokensharing/TokenShareUtility.java

```diff
@@ -0,0 +1,225 @@
+                    TAG + methodName,
+                    "That's strange, we had an AccountRecord for OID: "
+                            + oid
+                            + " but couldn't find tokens for them."
+            );
+
+            throw new ClientException(TOKEN_CACHE_ITEM_NOT_FOUND);
+        }
+
+        final TokenCacheItem cacheItemToExport = adapt(
+                cacheRecord.getIdToken(),
+                cacheRecord.getRefreshToken()
+        );
+
+        // Ship it
+        return SSOStateSerializer.serialize(cacheItemToExport);
+    }
+
+    @Override
+    public void saveFamilyRefreshToken(@NonNull final String tokenCacheItemJson) throws Exception {
+        final Future<Pair<MicrosoftAccount, MicrosoftRefreshToken>> resultFuture =
+                sBackgroundExecutor.submit(new Callable<Pair<MicrosoftAccount, MicrosoftRefreshToken>>() {
+                    @Override
+                    public Pair<MicrosoftAccount, MicrosoftRefreshToken> call() throws ClientException {
+                        final TokenCacheItem cacheItemToRenew = SSOStateSerializer.deserialize(tokenCacheItemJson);
+
+                        // We're going to 'hijack' this token and set our own clientId for renewal
+                        // since these are FRTs, this is OK to do.
+                        cacheItemToRenew.setClientId(mClientId);
+
+                        return renewToken(mRedirectUri, cacheItemToRenew);
+                    }
+                });
+
+        final Pair<MicrosoftAccount, MicrosoftRefreshToken> resultPair = resultFuture.get();
+
+        // If an error is encountered while requesting new tokens, null is returned
+        // Check the result, before proceeding to save into the cache...
+        if (null != resultPair) {
+            mTokenCache.setSingleSignOnState(
+                    resultPair.first, // The account
+                    resultPair.second // The refresh token
+            );
+        }
+    }
+
+    @SuppressWarnings("PMD.UnusedPrivateMethod")
+    @NonNull
+    private static TokenCacheItem adapt(@NonNull final IdTokenRecord idTokenRecord,
+                                        @NonNull final RefreshTokenRecord refreshTokenRecord) throws ServiceException {
+        final TokenCacheItem tokenCacheItem = new TokenCacheItem();
+        tokenCacheItem.setClientId(refreshTokenRecord.getClientId());
+        tokenCacheItem.setAuthority(idTokenRecord.getAuthority());
+        tokenCacheItem.setRefreshToken(refreshTokenRecord.getSecret());
+        tokenCacheItem.setResource(null); // TODO Does this need to be present? You can't turn scopes into a resource AFAIK...
+        tokenCacheItem.setRawIdToken(mintV1IdToken(idTokenRecord.getSecret()));
+        tokenCacheItem.setFamilyClientId(refreshTokenRecord.getFamilyId());
+
+        return tokenCacheItem;
+    }
+
+    @NonNull
+    private static String mintV1IdToken(@NonNull final String rawV2IdToken) throws ServiceException {
```

Comment:
We should specify in the name of the method that we're creating/minting a v1 token from a v2 id_token

Reply:
Resolved in 042e9639d273e518ded148bd55cb2848c155c12a